File: app/src/lib/categories.ts
// Pure category-normalization module (feature 002, issue #12).
// Merges near-duplicate AI-chosen labels so board sections don't split.

const collapse = (raw: string): string => raw.trim().replace(/\s+/g, ' ');

const fold = (raw: string): string =>
  collapse(raw.replace(/[-_]/g, ' ')).toLowerCase();

// Trailing-s plural fold on the last word; words shorter than 4 chars are left
// alone so labels like "AI" or "iOS" are never mangled.
const singularFold = (folded: string): string => {
  const words = folded.split(' ');
  const last = words[words.length - 1];
  if (last !== undefined && last.length >= 4 && last.endsWith('s')) {
    words[words.length - 1] = last.slice(0, -1);
  }
  return words.join(' ');
};

const initials = (folded: string): string | null => {
  const words = folded.split(' ');
  if (words.length < 2) return null;
  return words.map((w) => w.charAt(0)).join('');
};

const levenshtein = (a: string, b: string): number => {
  let prev = Array.from({ length: b.length + 1 }, (_, i) => i);
  for (let i = 1; i <= a.length; i++) {
    const curr = [i];
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min((curr[j - 1] ?? 0) + 1, (prev[j] ?? 0) + 1, (prev[j - 1] ?? 0) + cost);
    }
    prev = curr;
  }
  return prev[b.length] ?? 0;
};

const typoMatches = (a: string, b: string): boolean => {
  if (a.length < 5 || b.length < 5) return false;
  const threshold = Math.min(a.length, b.length) >= 8 ? 2 : 1;
  return levenshtein(a, b) <= threshold;
};

/**
 * Given an AI-proposed category label and the labels already on the board,
 * returns the label to file the card under: an existing label when the new
 * one is a near-duplicate (case/whitespace, plural, acronym, single typo),
 * otherwise the cleaned-up new label.
 */
export const resolveCategory = (raw: string, existing: string[]): string => {
  const cleaned = collapse(raw);
  if (cleaned === '') return 'Uncategorized';
  const f = fold(raw);

  const rules: Array<(e: string) => boolean> = [
    (e) => fold(e) === f,
    (e) => singularFold(fold(e)) === singularFold(f),
    (e) => initials(fold(e)) === f || initials(f) === fold(e),
    (e) => typoMatches(fold(e), f),
  ];
  for (const matches of rules) {
    const hit = existing.find(matches);
    if (hit !== undefined) return hit;
  }
  return cleaned;
};

File: api/categories.py
"""Category normalization on the save path (feature 020, issue #12).

Faithful Python port of app/src/lib/categories.ts (feature 002): merges
near-duplicate AI-chosen labels into existing sections. Rules and priority
order are identical; test_categories.py mirrors categories.spec.ts to keep
the two implementations in sync.
"""

from __future__ import annotations

import re


def _collapse(raw: str) -> str:
    return re.sub(r"\s+", " ", raw.strip())


def _fold(raw: str) -> str:
    return _collapse(re.sub(r"[-_]", " ", raw)).lower()


def _singular_fold(folded: str) -> str:
    """Trailing-s plural fold on the last word; words shorter than 4 chars are
    left alone so labels like "AI" or "iOS" are never mangled."""
    words = folded.split(" ")
    last = words[-1]
    if len(last) >= 4 and last.endswith("s"):
        words[-1] = last[:-1]
    return " ".join(words)


def _initials(folded: str) -> str | None:
    words = folded.split(" ")
    if len(words) < 2:
        return None
    return "".join(word[0] for word in words)


def _levenshtein(a: str, b: str) -> int:
    prev = list(range(len(b) + 1))
    for i in range(1, len(a) + 1):
        curr = [i]
        for j in range(1, len(b) + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            curr.append(min(curr[j - 1] + 1, prev[j] + 1, prev[j - 1] + cost))
        prev = curr
    return prev[len(b)]


def _typo_matches(a: str, b: str) -> bool:
    if len(a) < 5 or len(b) < 5:
        return False
    threshold = 2 if min(len(a), len(b)) >= 8 else 1
    return _levenshtein(a, b) <= threshold


def resolve_category(raw: str, existing: list[str]) -> str:
    """Given an AI-proposed category label and the labels already stored,
    returns the label to file the card under: an existing label when the new
    one is a near-duplicate (case/whitespace, plural, acronym, single typo),
    otherwise the cleaned-up new label."""
    cleaned = _collapse(raw)
    if cleaned == "":
        return "Uncategorized"
    folded = _fold(raw)

    rules = [
        lambda e: _fold(e) == folded,
        lambda e: _singular_fold(_fold(e)) == _singular_fold(folded),
        lambda e: _initials(_fold(e)) == folded or _initials(folded) == _fold(e),
        lambda e: _typo_matches(_fold(e), folded),
    ]
    for matches in rules:
        hit = next((e for e in existing if matches(e)), None)
        if hit is not None:
            return hit
    return cleaned
